src/components/Login/TermsConditionsAcceptButton.tsx

```diff
@@ -30,89 +22,84 @@ type OwnProps = {
-
-const TermsConditionsAcceptButton: FunctionComponent<Props> = ({
-  isNewUser,
-  disabled,
-  history,
-  state: { loggedInUser, isLoadingUser },
-  effects: { setUser },
-  onSubscribeUser,
-  api,
-  type,
-}) => {
-  const [isUpdatingUser, setIsUpdatingUser] = useState(false);
-  const onClick = async () => {
-    setIsUpdatingUser(true);
-    const { email, ...rest } = loggedInUser;
-    // FIXME Legacy code
-    const profile = await updateProfile(api)({
-      user: {
-        ...rest,
-        email,
-        acceptedTerms: true,
-      },
+type Props = OwnProps & PropsFromRedux;
+
+const TermsConditionsAcceptButton = ({ isNewUser, onSubscribeUser, user, acceptTerms }: Props) => {
+  const history = useHistory();
+  const [updating, setUpdating] = useState(false);
+
+  const onFinish = async () => {
+    setForumBannerForDisplay();
+    if (!isNewUser) {
+      setUpdating(false);
+      history.push(ROUTES.dashboard);
+      return;
+    }
+
+    onSubscribeUser();
+
+    notification.info({
+      message: 'Success',
+      description: (
+        <Paragraph>
+          Fill out your profile, or skip and <a href={'/search/file'}>browse data</a>
+        </Paragraph>
+      ),
+      duration: NEVER_CLOSES_AUTOMATICALLY,
+      key: 'terms-conditions-info-notification',
+    });
+
+    await trackUserInteraction({
+      value: undefined,
+      category: TRACKING_EVENTS.categories.join,
+      action: TRACKING_EVENTS.actions.signedUp,
+      label: `Join Completion: egoId ${user!._id}`,
     });
+    setUpdating(false);
+    history.push(`${ROUTES.user}/${user!._id}`);
+  };
 
-    const updatedUser = { ...profile, email };
-    setIsUpdatingUser(false);
-    if (isNewUser) {
-      await onSubscribeUser(updatedUser);
-      notification.info({
-        message: 'Success',
-        description: (
-          <Paragraph>
-            Fill out your profile, or skip and <a href={'/search/file'}>browse data</a>
-          </Paragraph>
-        ),
-        duration: NEVER_CLOSES_AUTOMATICALLY,
-        key: 'terms-conditions-info-notification',
-      });
-      await trackUserInteraction({
-        value: undefined,
-        category: TRACKING_EVENTS.categories.join,
-        action: TRACKING_EVENTS.actions.signedUp,
-        label: `Join Completion: egoId ${loggedInUser._id}`,
-      });
-      await setUser({ ...updatedUser, api, isJoining: false });
-      setForumBannerForDisplay();
-      history.push(`${ROUTES.user}/${loggedInUser._id}`);
-    } else {
-      await setUser({ ...updatedUser, api, isJoining: false });
-      setForumBannerForDisplay();
+  const onClick = async () => {
+    if (user && user.acceptedTerms) {
+      // Page on unprotected route.
+      // case: someone is logged in + already accepted terms
+      // he/she wanted to decline terms but changed his/her mind.
+      // Then, no need to update anything.
       history.push(ROUTES.dashboard);
+      return;
```

Comment:
components should return null, not undefined

Reply:
it'is an onClick handler embedeed in a component.